### PR TITLE
REDSHOP-6131 Automation - [New feature] Promotion

### PR DESCRIPTION
**Pull request by NguyenTrang091296**

### CODE REVIEW

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+					if(isset($promotion['freeShipping']))
+					{
+						$I->waitForElementVisible(PromotionsPage::$selectFreeShipping, 30);
+						$I->click(PromotionsPage::$selectFreeShipping);
+						$I->waitForElementVisible(PromotionsPage::$searchFreeShipping, 30);
+						$I->fillField(PromotionsPage::$searchFreeShipping, $promotion['freeShipping']);
+						$I->pressKey(PromotionsPage::$searchFreeShipping, \Facebook\WebDriver\WebDriverKeys::ENTER);
+					}
+				}
+			}else
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space line before "else"

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+		$I = $this;
+		$I->doFrontEndLogin($customerInformation['userName'], $customerInformation['password']);
+
+		if($promotion['promotionType'] == 'Amount Product')
+		{
+			for( $a= 0; $a < $promotion['conditionAmount']; $a++)
+			{
+				$I->addToCartAjax($promotion['category'], $promotion['product'], 'no', 'no');
+			}
+		}else
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add spaceline before "else"

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	 * @param $customerInformation
+	 * @throws Exception
+	 * @since 3.0.3
+	 */
+	public function checkPromotionWithCartAjax($promotion, $customerInformation)
+	{
+		$I = $this;
+		$I->doFrontEndLogin($customerInformation['userName'], $customerInformation['password']);
+
+		if($promotion['promotionType'] == 'Amount Product')
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and  "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+		$I->click(FrontEndProductManagerJoomla3Page::$checkoutFinalStep);
+		$I->waitForElementVisible(FrontEndProductManagerJoomla3Page::$orderReceiptTitle, 30);
+
+		if($promotion['freeShipping'] == 'No')
+		{
+			$priceRate = $currencyUnit['currencySymbol'].($shipping['shippingRate']).$currencyUnit['decimalSeparator'].$currencyUnit['numberZero'];
+			$I->waitForText($priceRate, 30);
+			$I->see($priceRate);
+			$I->doFrontendLogout();
+		}else
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add spaceline before "else"

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+		$I->wait(0.5);
+		$I->waitForElementVisible(FrontEndProductManagerJoomla3Page::$termAndConditions, 60);
+		$I->seeElement(FrontEndProductManagerJoomla3Page::$termAndConditions);
+		$I->click(FrontEndProductManagerJoomla3Page::$termAndConditions);
+		$I->waitForElementVisible(FrontEndProductManagerJoomla3Page::$checkoutFinalStep, 30);
+		$I->seeElement(FrontEndProductManagerJoomla3Page::$checkoutFinalStep);
+		$I->click(FrontEndProductManagerJoomla3Page::$checkoutFinalStep);
+		$I->waitForElementVisible(FrontEndProductManagerJoomla3Page::$orderReceiptTitle, 30);
+
+		if($promotion['freeShipping'] == 'No')
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+	 * @throws Exception
+	 * @since 3.0.3
+	 */
+	public function checkoutPromotionWithShipping($promotion, $shipping, $customerInformation)
+	{
+		$I = $this;
+		$currencyUnit = $I->getCurrencyValue();
+		$I->doFrontEndLogin($customerInformation['userName'], $customerInformation['password']);
+
+		if($promotion['promotionType'] == 'Amount Product')
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+						$I->pressKey(PromotionsPage::$searchFreeShipping2, \Facebook\WebDriver\WebDriverKeys::ENTER);
+					}
+				}
+			}
+		}
+
+		$I->waitForElementVisible(PromotionsPage::$idFieldName, 30);
+		$I->fillField(PromotionsPage::$idFieldName, $promotion['name']);
+
+		if(isset($promotion['desc']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+					$I->fillField(PromotionsPage::$searchProductAwards2, $promotion['productAward']);
+					$I->pressKey(PromotionsPage::$searchProductAwards2, \Facebook\WebDriver\WebDriverKeys::ENTER);
+
+					if(isset($promotion['awardAmount']))
+					{
+						$I->waitForElementVisible(PromotionsPage::$inputAwardAmount, 30);
+						$I->fillField(PromotionsPage::$inputAwardAmount, $promotion['awardAmount']);
+					}
+
+					if(isset($promotion['freeShipping']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+				if(isset($promotion['productAward']))
+				{
+					$I->waitForElementVisible(PromotionsPage::$selectProductAwards, 30);
+					$I->click(PromotionsPage::$selectProductAwards);
+					$I->waitForElementVisible(PromotionsPage::$searchProductAwards2, 30);
+					$I->fillField(PromotionsPage::$searchProductAwards2, $promotion['productAward']);
+					$I->pressKey(PromotionsPage::$searchProductAwards2, \Facebook\WebDriver\WebDriverKeys::ENTER);
+
+					if(isset($promotion['awardAmount']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+				$I->waitForElementVisible(PromotionsPage::$inputOrderVolume, 30);
+				$I->fillField(PromotionsPage::$inputOrderVolume, $promotion['orderVolume']);
+
+				$I->waitForElementVisible(PromotionsPage::$inputFromDate, 30);
+				$I->fillField(PromotionsPage::$inputFromDate, $promotion['fromDate']);
+
+				$I->waitForElementVisible(PromotionsPage::$inputToDate, 30);
+				$I->fillField(PromotionsPage::$inputToDate, $promotion['toDate']);
+
+				if(isset($promotion['productAward']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+					$I->fillField(PromotionsPage::$searchProductAwards, $promotion['productAward']);
+					$I->pressKey(PromotionsPage::$searchProductAwards, \Facebook\WebDriver\WebDriverKeys::ENTER);
+
+					if(isset($promotion['awardAmount']))
+					{
+						$I->waitForElementVisible(PromotionsPage::$inputAwardAmount, 30);
+						$I->fillField(PromotionsPage::$inputAwardAmount, $promotion['awardAmount']);
+					}
+
+					if(isset($promotion['freeShipping']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+				$I->waitForElementVisible(PromotionsPage::$inputConditionAmount, 30);
+				$I->fillField(PromotionsPage::$inputConditionAmount, $promotion['conditionAmount']);
+
+				$I->waitForElementVisible(PromotionsPage::$inputFromDate, 30);
+				$I->fillField(PromotionsPage::$inputFromDate, $promotion['fromDate']);
+
+				$I->waitForElementVisible(PromotionsPage::$inputToDate, 30);
+				$I->fillField(PromotionsPage::$inputToDate, $promotion['toDate']);
+
+				if(isset($promotion['productAward']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+				{
+					$I->waitForElementVisible(PromotionsPage::$selectProductAwards, 30);
+					$I->click(PromotionsPage::$selectProductAwards);
+					$I->waitForElementVisible(PromotionsPage::$searchProductAwards, 30);
+					$I->fillField(PromotionsPage::$searchProductAwards, $promotion['productAward']);
+					$I->pressKey(PromotionsPage::$searchProductAwards, \Facebook\WebDriver\WebDriverKeys::ENTER);
+
+					if(isset($promotion['awardAmount']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+
+		if(isset($promotion['promotionType']))
+		{
+			$I->waitForElementVisible(PromotionsPage::$selectPromotionType, 30);
+			$I->click(PromotionsPage::$selectPromotionType);
+			$I->waitForElementVisible(PromotionsPage::$searchPromotionType, 30);
+			$I->fillField(PromotionsPage::$searchPromotionType, $promotion['promotionType']);
+			$I->pressKey(PromotionsPage::$searchPromotionType, \Facebook\WebDriver\WebDriverKeys::ENTER);
+
+			if($promotion['promotionType'] == 'Amount Product')
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+	 */
+	public function createPromotion($promotion, $function)
+	{
+		$I = $this;
+		$I->amOnPage(PromotionsPage::$url);
+		$I->waitForText(PromotionsPage::$titlePage, 30, PromotionsPage::$h1);
+		$I->click(PromotionsPage::$buttonNew);
+		$I->waitForText(PromotionsPage::$titlePageNew, 30, PromotionsPage::$h1);
+
+		if(isset($promotion['promotionType']))
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space between if and "("

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+
+		$I->waitForElementVisible(PromotionsPage::$idFieldName, 30);
+		$I->fillField(PromotionsPage::$idFieldName, $promotion['name']);
+
+		if(isset($promotion['desc']))
+		{
+			$I->waitForElementVisible(PromotionsPage::$textareDescription, 30);
+			$I->fillField(PromotionsPage::$textareDescription, $promotion['desc']);
+		}
+		Switch ($function)
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php:135]: add space line before "switch"

File: tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+			$I->click(["link" => $promotion['product']]);
+
+			for( $a= 0; $a < $promotion['conditionAmount']; $a++)
+			{
+				$I->waitForElementVisible(AdminJ3Page::$addToCart, 30);
+				$I->seeElement(AdminJ3Page:: $addToCart);
+				$I->click(AdminJ3Page:: $addToCart);
+				$I->waitForText(AdminJ3Page::$alertSuccessMessage, 120, FrontEndProductManagerJoomla3Page::$selectorSuccess);
+			}
+		}else
```
Comment (VanTrieuThanh) [inline tests/acceptance/administrator/_steps/Joomla3Steps/PromotionsManagementSteps.php]: add space line before "else"

**Review by VanTrieuThanh — COMMENTED** [review]
Let's review my comment

**Review by ghost — DISMISSED** [review]

**Review by VanTrieuThanh — DISMISSED** [review]

**Review by VanTrieuThanh — APPROVED** [review]

**Pull request merged**